File: modules/libs/core/appearance/appearance.go
// Package appearance is the settings' appearance section as a page carries it.
//
// Every window of the installation shows the same one: the mode, the theme's
// own file, and the two sizes it is drawn and set at. A person sets those once
// and every application they open shows them, so how they are written into a
// page is one thing here and not one per window.
package appearance

import (
	"bytes"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

// HeadEnd is where the style elements go: after everything the build put in the
// head, so what a person chose is the last word on it.
const HeadEnd = "</head>"

// Marker is the attribute every one of those elements carries, and Mode, Theme
// and Sizes are what each says it is. A window reads them by it.
const (
	Marker  = "data-appearance"
	IsMode  = "mode"
	IsTheme = "theme"
	IsSizes = "sizes"
)

// ColorScheme is which half of every colour pair the tokens are read as.
type ColorScheme int

const (
	// System is both halves, and the machine decides between them.
	System ColorScheme = iota
	Light
	Dark
)

// Settings is what a person set their windows to.
type Settings struct {
	Mode ColorScheme
	// Theme is the stylesheet chosen, as its text. A window wearing none is
	// left to `tokens.css`.
	Theme string
	// InterfaceScale is how large the interface is drawn and TextScale how
	// large the text a person reads is set, one being as designed. A size
	// nobody named is zero and is left to `tokens.css` as well.
	InterfaceScale float64
	TextScale      float64
}

// Styles is the elements the head ends with.
//
// The page arrives carrying all of them, so no frame is drawn in the default
// colours or at a size nobody asked for. A size arriving after the first frame
// relays out the document.
func Styles(c Settings) string {
	// The mode first and the theme second. A theme pinning `color-scheme` is
	// the later of two declarations weighing the same, and light and dark are
	// then that theme's own.
	out := styled(IsMode, ":root { color-scheme: "+scheme(c.Mode)+"; }")
	if c.Theme != "" {
		out += styled(IsTheme, c.Theme)
	}
	// The two sizes last. They are what a person set this window to, inside the
	// bounds each goes to, and the element carrying them is the last word on
	// them.
	return out + sized(c.InterfaceScale, c.TextScale)
}

// Into is the page carrying those elements, put where the head ends. A page
// with no head is handed back as it was built.
func Into(text []byte, styles string) []byte {
	at := bytes.LastIndex(text, []byte(HeadEnd))
	if styles == "" || at < 0 {
		return text
	}
	return slices.Concat(text[:at], []byte(styles), text[at:])
}

// sized is the two multipliers as the page carries them: how large the
// interface is drawn, which is the root's font size, and how large the text a
// person reads is set.
func sized(drawn, set float64) string {
	var held []string
	if drawn > 0 {
		held = append(held, "--numen-interface-scale: "+number(drawn))
	}
	if set > 0 {
		held = append(held, "--numen-text-scale: "+number(set))
	}
	if len(held) == 0 {
		return ""
	}
	return styled(IsSizes, ":root { "+strings.Join(held, "; ")+"; }")
}

// number is a multiplier as CSS takes it, at the shortest that reads back as
// the number it was given.
func number(size float64) string { return strconv.FormatFloat(size, 'f', -1, 64) }

// scheme is which half of every colour pair the tokens are read as.
func scheme(mode ColorScheme) string {
	switch mode {
	case Light:
		return "light"
	case Dark:
		return "dark"
	default:
		return "light dark"
	}
}

// styled is one stylesheet as the page carries it, marked as the one of the
// three it is. A `</style>` in a theme's file is written as the CSS escape for
// it, which is the same declaration and ends no element.
func styled(is, css string) string {
	escaped := ending.ReplaceAllStringFunc(css, func(found string) string {
		return `<\` + found[len("<"):]
	})
	return `<style ` + Marker + `="` + is + `">` + escaped + "</style>\n"
}

var ending = regexp.MustCompile(`(?i)</style`)

File: modules/apps/desktop/editor/src/features/settings-commands/model/appearance.ts
/**
 * The window as it is drawn — the theme it wears, which half of a colour pair
 * its tokens are read as, and how large it is drawn and its reading text set.
 *
 * A theme and a mode are worn the moment the keyboard lands on them. A size is
 * held until the keyboard has stood on the row for HELD, and what the settings
 * name is put back the moment the keyboard leaves the list.
 */
import { ref, shallowRef, watch } from 'vue'
import { createFollower } from '@numen/ui'
import { answerGuard } from '@/shared/questions'
import type { MessageWriter } from '@/shared/notices/messages'
import { DESIGNED, NOWHERE, SCHEMES } from '@/entities/settings'
import type { Appearance, Mode, Ranges, Sizes, Theme, Themes } from '@/entities/settings'
import type { AppearanceWords } from '../words'
import { IS_SIZES, after, getSizesCss, getStyleElements } from '../lib/head'
import { getModeGroups, getSizeGroups, getThemeGroups } from '../lib/offers'
import { useAppearanceChoice } from './choice'

export function windowAppearance(
  core: Themes,
  words: AppearanceWords,
  write: MessageWriter,
  sheet: Document = document,
  wait: (ms: number) => Promise<unknown> = sleep,
) {
  const dressed = getStyleElements(sheet)
  /** What the page was served wearing, which is the applied theme's file. */
  const served = dressed.theme.textContent ?? ''
  /** What the sizes' element holds, once the window knows what it was served at. */
  let written = ''

  /** Every theme there is, and the theme and the mode the settings name. */
  const list = shallowRef<readonly Theme[]>([])
  const applied = ref('')
  const mode = ref<Mode>('system')

  /** The two sizes the settings name, and how far each of them goes. */
  const settings = ref<Sizes>({ interfaceScale: DESIGNED, textScale: DESIGNED })
  const bounds = ref<Ranges>({ interfaceScale: NOWHERE, textScale: NOWHERE })

  /**
   * What the appearance lost touch with, said until it has it back.
   *
   * It is a state and not a word, so the window carries it where it carries
   * everything else that is simply so.
   */
  const lost = ref('')

  /** The text of every theme that has been worn, by name. */
  const files = new Map<string, string>()

  /** Whether what the page arrived in still stands for the theme applied. */
  let arrived = true

  /** A file arriving for a row the keyboard has already left is dropped. */
  const asks = answerGuard()

  /** One theme's file, read once and kept. */
  const fileOf = async (name: string): Promise<string> => {
    const kept = files.get(name)
    if (kept !== undefined) return kept
    const css = await core.readTheme(name)
    files.set(name, css)
    return css
  }

  /** The theme and the mode named, written into the two elements the page was served with. */
  const wear = async (theme: string, half: Mode) => {
    // The page arrived dressed, and nothing is written over that until the
    // window has been told what it is dressed in.
    if (!applied.value) return
    const mine = asks.ask()
    dressed.mode.textContent = `:root { color-scheme: ${SCHEMES[half]}; }`
    try {
      const css = await fileOf(theme)
      if (mine.current) dressed.theme.textContent = css
    } catch (error) {
      // The reason goes to the console; the person is told in the window's
      // own voice.
      console.error(error)
      if (mine.current) write(words.unworn, 'error')
    }
  }

  /** The row the keyboard stands on, and what standing on it or choosing it does. */
  const choice = useAppearanceChoice({
    core,
    write,
    list,
    applied,
    mode,
    settings,
    bounds,
    wear,
  })

  /** What the window wears now, which is the row stood on over the settings. */
  const applyAppearance = () => wear(choice.worn.value, choice.half.value)

  /**
   * The two multipliers, written into the element the head ends with. The page
   * was served carrying them, so what already stands there is left alone.
   */
  const writeSizes = () => {
    const css = getSizesCss(choice.sized.value)
    if (css === written) return
    written = css
    dressed.sizes ??= after(dressed.theme, IS_SIZES, sheet)
    dressed.sizes.textContent = css
  }

  /** The window drawn again wherever the size it is drawn at changes. */
  const drawing = watch(choice.sized, writeSizes)

  let open = true
  /** Let go of the stream the window is listening to. */
  const listening = new AbortController()
  const follows = createFollower({
    isOpen: () => open,
    setLost: (gone) => (lost.value = gone),
    wait,
  })

  /** Every theme there is, and what the settings say the window is drawn as. */
  const loadAppearance = async () => {
    let answer: Appearance
    try {
      answer = await core.getAppearance()
    } catch (error) {
      console.error(error)
      write(words.unlisted, 'error')
      return
    }
    list.value = answer.themes
    applied.value = answer.applied
    mode.value = answer.mode
    settings.value = answer.sizes
    bounds.value = answer.bounds
    if (arrived) {
      // The page was served wearing this theme, so its file has been read
      // already, and drawn at these sizes, so they already stand in the head.
      files.set(answer.applied, served)
      written = dressed.sizes?.textContent ?? getSizesCss(answer.sizes)
    }
    arrived = false
  }

  /**
   * The person's folder changed. What changed is read again from the file it
   * is now, and a theme being worn is worn as it now reads.
   */
  const follow = () =>
    follows(
      () => core.watchThemes(listening.signal),
      async (names) => {
        if (!names.length) return
        for (const name of names) files.delete(name)
        await loadAppearance()
        if (names.includes(choice.worn.value)) await applyAppearance()
      },
    )

  /**
   * The list, and the folder followed. Nothing is worn from here: the page
   * arrived dressed, and this is what it was dressed in.
   */
  const start = async () => {
    await loadAppearance()
    void follow()
  }

  const close = () => {
    open = false
    listening.abort()
    choice.close()
    drawing()
  }

  const listThemeGroups = () => getThemeGroups(list.value, applied.value, words)
  const listModeGroups = () => getModeGroups(mode.value, choice.isPinned.value, words)
  const listSizeGroups = (command: string, text = '') =>
    getSizeGroups(command, text, settings.value, bounds.value, words)

  return {
    list,
    applied,
    mode,
    sized: choice.sized,
    bounds,
    isPinned: choice.isPinned,
    pinned: choice.isPinned,
    lost,
    getThemeGroups: listThemeGroups,
    getModeGroups: listModeGroups,
    getSizeGroups: listSizeGroups,
    previewItem: choice.previewItem,
    chooseItem: choice.chooseItem,
    start,
    close,
  }
}

const sleep = (ms: number) => new Promise((wake) => setTimeout(wake, ms))
